Fix assembler detection in configure

shindere opened the pull request:
This PR fixes the detection of the assembler program on some platforms,
amnong which Mac OS. It thus fixes MPR#7919.

The PR also makes sure the CPP build variable is dealt with properly,
which should also be helpful for GPR##2295

See precheck build #216(https://ci.inria.fr/ocaml/job/precheck/216).

(Should be cherry-picked to 4.08.)

shindere committed:
Clarify the invocation of the C preprocessor

Most of the time, the C preprocessor needs to be invoked through the C
compiler, e.g. so that the paths to the header files are resolved properly.
In some cases, though, we really need to be able to call the C
preprocessor directly, just to expand macros in .ml files (this only
happens in the testsuite, at the moment). In those cases, it is
simply impossible to call the C preprocessor through the compiler,
e.g. because this would require the input files to have a '.c'
extension, which the OCaml compiler would misinterprete as meaning this file
should be compiled with the C compiler.

Thus, this commit clarifies the distinction between CPP and DIRECT_CPP
and provides both variables to the build system. The ocamltest build system
is also updated to take advantage of this.

We rely on autoconf's macros to detect how to call the C preprocessor
via the C compiler, except for the MSVC port where its value is hard-coded
to guarantee backward compatibility.

shindere committed:
Fix detection of assembler on some platforms

This commit fixes MPR#7919

shindere committed:
Update Changes

dra27 commented on configure.ac:
```diff
@@ -73,6 +73,7 @@ AC_SUBST([CC])
 # Note: This is present for the flexdll bootstrap where it exposed as the old
 # TOOLPREF variable. It would be better if flexdll where updated to require
 # WINDRES instead.
+AC_SUBST([DIRECT_CPP])
```
Hmm, `AC_CHECK_TOOL` already calls `AC_SUBST`, but of course `AC_CHECK_TOOL` is not being called on MSVC. Perhaps it would be better to be explicit further down, and have `AC_SUBST([DIRECT_CPP],[$CPP])` instead of my suggested `DIRECT_CPP="$CPP"`?

dra27 commented on configure.ac:
```diff
@@ -403,12 +404,19 @@ AS_CASE([$host],
 AS_CASE([$host],
-  [*-pc-windows], [CPP="$CC -nologo -EP"],
-  [CPP=$DIRECT_CPP])
+  [*-pc-windows], [DIRECT_CPP="$CC -nologo -EP"],
```
Following on from our private discussion, autoconf will infer `CPP=cl -E` which is different (`-EP` suppresses `#line` directives and `-nologo` suppresses the Microsoft copyright notice). So I think the MSVC part needs to stay as it is but add `DIRECT_CPP="$CPP"`

dra27 reviewed (commented):
Tweak needed for MSVC - I'm not sure how important it actually is, but it seems sensible to maintain the previous value of `CPP` on the two MSVC ports.

shindere commented:
David Allsopp (2019/03/12 04:28 -0700):
> dra27 commented on this pull request.
> 
> Tweak needed for MSVC - I'm not sure how important it actually is, but
> it seems sensible to maintain the previous value of `CPP` on the two
> MSVC ports.

Yes, you are right of course! Fixed.



> Hmm, `AC_CHECK_TOOL` already calls `AC_SUBST`, but of course
> `AC_CHECK_TOOL` is not being called on MSVC. Perhaps it would be
> better to be explicit further down, and have
> `AC_SUBST([DIRECT_CPP],[$CPP])` instead of my suggested
> `DIRECT_CPP="$CPP"`?

Could you please have a look at how I modified the code and tell me what
you think about this new version? For the moment I kept the call to
AC_SUBST because it felt more explicit to me, what do you think?

> >  AS_CASE([$host],
> -  [*-pc-windows], [CPP="$CC -nologo -EP"],
> -  [CPP=$DIRECT_CPP])
> +  [*-pc-windows], [DIRECT_CPP="$CC -nologo -EP"],
> 
> Following on from our private discussion, autoconf will infer `CPP=cl
> -E` which is different (`-EP` suppresses `#line` directives and
> `-nologo` suppresses the Microsoft copyright notice). So I think the
> MSVC part needs to stay as it is but add `DIRECT_CPP="$CPP"`

I implemented this.*

Let me know if that looks good.

dra27 commented:
@shindere wrote:
> Could you please have a look at how I modified the code and tell me what
> you think about this new version? For the moment I kept the call to
> AC_SUBST because it felt more explicit to me, what do you think?

I was suggesting to keep the `AC_SUBST`, just to put it in a different place - what "worries" me slightly is that I saw it and thought "it's not necessary, that variable's initialised with `AC_CHECK_TOOL`", but that's not actually true for MSVC. So by having `AC_SUBST([DIRECT_CPP],[$CPP])` later, that felt to me more explicit.

ksromanov commented:
I rebased changes in GPR #2295 on top of this branch - everything works exceptionally well!

Tested systems (compile, run repl, `#load "unix.cma"`):

1. xlc/AIX/POWER/32-bit
2. xlc/AIX/POWER/64-bit
3. gcc4.3/AIX/POWER/32-bit
4. gcc5.3/linux/x86-32

shindere commented:
David Allsopp (2019/03/12 08:40 -0700):
> @shindere wrote:
> > Could you please have a look at how I modified the code and tell me what
> > you think about this new version? For the moment I kept the call to
> > AC_SUBST because it felt more explicit to me, what do you think?
> 
> I was suggesting to keep the `AC_SUBST`, just to put it in a different
> place - what "worries" me slightly is that I saw it and thought "it's
> not necessary, that variable's initialised with `AC_CHECK_TOOL`", but
> that's not actually true for MSVC. So by having
> `AC_SUBST([DIRECT_CPP],[$CPP])` later, that felt to me more explicit.

So, do you want me to change it, or is it okay as it is?

shindere commented:
Sounds good! Thanks!

dra27 commented:
@shindere - for the `AC_SUBST` change, that's up-to-you! I've stated what worries me about how it is, which you're very welcome to either agree with and change to, or not share my worry :) I don't insist, no - especially as I can see why you grouped all the `AC_SUBST` lines up there in the first place.

dra27 commented:
There will presumably be another beta before 4.08 goes out of the door, @damiendoligez ... minor though the change hopefully is, the effective change to `utils/config.ml` by going with autoconf's `CPP` value I don't think should appear between beta and release?

dra27 commented:
Other than that, LGTM.

shindere commented:
David Allsopp (2019/03/12 09:27 -0700):
> @shindere - for the `AC_SUBST` change, that's up-to-you! I've stated
> what worries me about how it is, which you're very welcome to either
> agree with and change to, or not share my worry :) I don't insist, no
> - especially as I can see why you grouped all the `AC_SUBST` lines up
> there in the first place.

@dra27 thanks!

Then if you don't mind I'll let things as they are, at the moment. Not
so much becuase I don'et agree with the change or don't like it, but
rather because of tight schedule, sorry about that. If you wish to push
something on my branch you are very welcome to do so, though. Just make
sure to drp me a note because I am not sure I'd get any notification for
that.

shindere commented:
David Allsopp (2019/03/12 16:29 +0000):
> Other than that, LGTM.

Good, thanks.

So if you formally approve I could merge or you could do it, as you
prefer, so that we can move forward with #2295.

Re:4.08, are you saying that if no other beta is released, then this
should rather not be cherry-picked in its current state to the 4.08
branch, am I understanding you properly?

dra27 commented:
@shindere - re 4.08, yes. Happy to merge once CI comes back.

shindere commented:
David Allsopp (2019/03/12 09:36 -0700):
> @shindere - re 4.08, yes. Happy to merge once CI comes back.

Thanks @dra27! For that and for all your cooperation on this GPR and the
related stuff!

shindere commented:
CI is happy. could this please be approved?
Thanks!

shindere commented:
Thanks @dra27!

shindere commented:
@dra27 is it okay to cherry-pick to 4.08, finally?

dra27 commented:
@shindere - I just opened a PR for this, in case there's a delay deciding

shindere commented:
David Allsopp (2019/03/13 10:55 +0000):
> @shindere - I just opened a PR for this, in case there's a delay
> deciding

Many thanks for that, @dra27!
